File: 19_数学/矩阵运算/矩阵快速幂/例题/建信04. 电学实验课.go
// https://leetcode.cn/contest/ccbft-2021fall/problems/lSjqMF/
// # 实验目标要求同学们用导线连接所有「目标插孔」，
// # 即从任意一个「目标插孔」沿导线可以到达其他任意「目标插孔」
// # 一条导线可连接相邻两列的且行间距不超过 1 的两个插孔
// # 每一列插孔中最多使用其中一个插孔（包括「目标插孔」）
// # 若实验目标可达成，请返回使用导线数量最少的连接所有目标插孔的方案数；否则请返回 0。

// # 1 <= row <= 20
// # 3 <= col <= 10^9
// # 1 < position.length <= 1000

// # O(row^3*log(col)*position.length) = 20^3*log(10^9)*1000

package main

import (
	"sort"
)

const MOD int = 1e9 + 7

func electricityExperiment(row int, col int, position [][]int) int {
	n := len(position)
	sort.Slice(position, func(i, j int) bool {
		return position[i][1] < position[j][1]
	})

	T := make([][]int, row) // 转移矩阵
	for i := 0; i < row; i++ {
		T[i] = make([]int, row)
		T[i][i] = 1
		if i > 0 {
			T[i][i-1] = 1
		}
		if i < row-1 {
			T[i][i+1] = 1
		}
	}

	mp := NewMatPow(T, MOD, 4)
	res := 1
	for i := 0; i < n-1; i++ {
		r1, c1, r2, c2 := position[i][0], position[i][1], position[i+1][0], position[i+1][1]
		colDiff := c2 - c1
		pow_ := mp.Pow(colDiff)
		res = res * pow_[r1][r2] % MOD
	}
	return res
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

type M = [][]int

type MatPow struct {
	n          int
	mod        int
	base       []int
	cacheLevel int
	useCache   bool
	cache      [][][]int
}

// 带缓存的矩阵快速幂,适合多次查询
//
//	base: 转移矩阵,必须是方阵;
//	mod: 快速幂的模;
//	cacheLevel: 矩阵快速幂的log底数.启用缓存时一般设置为`4`;
//	当调用 pow 次数很多时,可设置为 `16`;
//	小于 `2` 时不启用缓存.
func NewMatPow(base M, mod, cacheLevel int) *MatPow {
	res := &MatPow{}
	n := len(base)
	base_ := make([]int, n*n)
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			base_[i*n+j] = base[i][j]
		}
	}
	useCache := cacheLevel >= 2

	res.n = n
	res.mod = mod
	res.base = base_
	res.cacheLevel = cacheLevel
	res.useCache = useCache
	if useCache {
		cache_ := make([][][]int, cacheLevel-1)
		for i := 0; i < cacheLevel-1; i++ {
			cache_[i] = make([][]int, 0, 64)
		}
		res.cache = cache_
	}

	return res
}

// 时间复杂度: O(n^3*log(exp))
func (mp *MatPow) Pow(exp int) M {
	if !mp.useCache {
		return mp.powWithOutCache(exp)
	}

	if len(mp.cache[0]) == 0 {
		mp.cache[0] = append(mp.cache[0], mp.base)
		for i := 1; i < mp.cacheLevel-1; i++ {
			mp.cache[i] = append(mp.cache[i], mp.mul(mp.cache[i-1][0], mp.base))
		}
	}

	e := mp.eye(mp.n)
	div := 0
	for exp > 0 {
		if div == len(mp.cache[0]) {
			mp.cache[0] = append(mp.cache[0], mp.mul(mp.cache[mp.cacheLevel-2][div-1], mp.cache[0][div-1]))
			for i := 1; i < mp.cacheLevel-1; i++ {
				mp.cache[i] = append(mp.cache[i], mp.mul(mp.cache[i-1][div], mp.cache[0][div]))
			}
		}

		mod := exp % mp.cacheLevel
		if mod > 0 {
			e = mp.mul(e, mp.cache[mod-1][div])
		}
		exp /= mp.cacheLevel
		div++
	}

	return mp.to2D(e)
}

func (mp *MatPow) mul(mat1, mat2 []int) []int {
	res := make([]int, mp.n*mp.n)
	for i := 0; i < mp.n; i++ {
		for k := 0; k < mp.n; k++ {
			for j := 0; j < mp.n; j++ {
				res[i*mp.n+j] = (res[i*mp.n+j] + mat1[i*mp.n+k]*mat2[k*mp.n+j]) % mp.mod
				if res[i*mp.n+j] < 0 {
					res[i*mp.n+j] += mp.mod
				}
			}
		}
	}
	return res
}

func (mp *MatPow) powWithOutCache(exp int) M {
	e := mp.eye(mp.n)
	b := append(mp.base[:0:0], mp.base...)
	for exp > 0 {
		if exp&1 == 1 {
			e = mp.mul(e, b)
		}
		b = mp.mul(b, b)
		exp >>= 1
	}

	return mp.to2D(e)
}

func (mp *MatPow) eye(n int) []int {
	res := make([]int, n*n)
	for i := 0; i < n; i++ {
		res[i*n+i] = 1
	}
	return res
}

func (mp *MatPow) to2D(mat []int) M {
	res := make(M, mp.n)
	for i := 0; i < mp.n; i++ {
		res[i] = make([]int, mp.n)
		copy(res[i], mat[i*mp.n:(i+1)*mp.n])
	}
	return res
}

File: 19_数学/矩阵运算/矩阵快速幂/例题/建信04. 电学实验课.js
// https://leetcode.cn/contest/ccbft-2021fall/problems/lSjqMF/
// # 实验目标要求同学们用导线连接所有「目标插孔」，
// # 即从任意一个「目标插孔」沿导线可以到达其他任意「目标插孔」
// # 一条导线可连接相邻两列的且行间距不超过 1 的两个插孔
// # 每一列插孔中最多使用其中一个插孔（包括「目标插孔」）
// # 若实验目标可达成，请返回使用导线数量最少的连接所有目标插孔的方案数；否则请返回 0。

// # 1 <= row <= 20
// # 3 <= col <= 10^9
// # 1 < position.length <= 1000

// # O(row^3*log(col)*position.length) = 20^3*log(10^9)*1000
// 超时

const MOD = 1e9 + 7

function electricityExperiment(row, col, position) {
  const T = Array.from({ length: row }, () => Array(row).fill(0))
  for (let r = 0; r < row; r++) {
    T[r][r] = 1
    if (r !== 0) T[r][r - 1] = 1
    if (r !== row - 1) T[r][r + 1] = 1
  }

  const n = position.length
  position.sort((a, b) => a[1] - b[1])
  for (let i = 0; i < n - 1; i++) {
    const [row1, col1] = position[i]
    const [row2, col2] = position[i + 1]
    const rowDiff = Math.abs(row1 - row2)
    const colDiff = Math.abs(col1 - col2)
    if (rowDiff > colDiff) return 0
  }

  let res = 1
  for (let i = 0; i < n - 1; i++) {
    const [row1, col1] = position[i]
    const [row2, col2] = position[i + 1]
    const colDiff = Math.abs(col1 - col2)
    res = mulUint32(res, cal(row1, row2, colDiff), MOD)
  }

  return Number(res)

  function cal(row1, row2, k) {
    const resT = matPow(T, k, MOD)
    return resT[row1][row2]
  }
}

function newMatrix(row, col) {
  const res = Array(row)
  for (let i = 0; i < row; i++) {
    res[i] = Array(col).fill(0)
  }
  return res
}

/**
 * 单位矩阵.
 * @example
 * ```ts
 * eye(3)
 * // [
 * //   [1, 0, 0],
 * //   [0, 1, 0],
 * //   [0, 0, 1]
 * // ]
 * ```
 */
function eye(n) {
  const res = Array(n)
  for (let i = 0; i < n; i++) {
    const row = Array(n).fill(0)
    row[i] = 1
    res[i] = row
  }
  return res
}

function copy(raw) {
  const row = raw.length
  const res = Array(row)
  for (let i = 0; i < row; i++) {
    res[i] = raw[i].slice()
  }
  return res
}

/**
 * uint32乘法.
 */
function mulUint32(num1, num2, mod = 1e9 + 7) {
  return (((Math.floor(num1 / 65536) * num2) % mod) * 65536 + (num1 & 65535) * num2) % mod
}

/**
 * uint32矩阵乘法.
 */
function matMul(mat1, mat2, mod = 1e9 + 7) {
  const row1 = mat1.length
  const row2 = mat2.length
  const col2 = mat2[0].length
  const res = newMatrix(row1, col2)
  for (let i = 0; i < row1; i++) {
    const resRow = res[i]
    const m1Row = mat1[i]
    for (let k = 0; k < row2; k++) {
      const m2Row = mat2[k]
      for (let j = 0; j < col2; j++) {
        resRow[j] = (resRow[j] + mulUint32(m1Row[k], m2Row[j], mod)) % mod
        if (resRow[j] < 0) resRow[j] += mod
      }
    }
  }
  return res
}

/**
 * uint32矩阵快速幂.
 */
function matPow(base, exp, mod = 1e9 + 7) {
  if (base.length !== base[0].length) throw new Error('base is not a square matrix')
  const n = base.length
  let res = eye(n)
  let baseCopy = copy(base)
  while (exp) {
    if (exp & 1) res = matMul(res, baseCopy, mod)
    baseCopy = matMul(baseCopy, baseCopy, mod)
    exp = Math.floor(exp / 2)
  }
  return res
}
